docs/ebpf/concepts/rlimit.md

```diff
@@ -0,0 +1,88 @@
+# Resource Limits
+
+Creating eBPF objects (Maps, Programs, even BTF blobs) requires kernel memory
+allocation. Before kernel version 5.11, you could control the memory available
+to a process for creating eBPF objects by adjusting the process'
+`RLIMIT_MEMLOCK` rlimit value, visible through the `ulimit -l` command.
+
+!!! note ""
+    User space typically does *not* allocate or `mlock(2)` memory for the
+    eBPF subsystem. The choice of `RLIMIT_MEMLOCK` was made early on due to its
+    connection to memory management. Unfortunately, this approach had
+    significant drawbacks, as explained in the [cover
+    letter](https://lore.kernel.org/bpf/20201201215900.3569844-1-guro@fb.com) of
+    the cgroup accounting introduced in 5.11.
+
+Starting with [version
+5.11](https://lore.kernel.org/bpf/20201201215900.3569844-1-guro@fb.com), the
+Linux kernel switched from rlimits to memory cgroup (memcg) accounting for
+managing memory limits on processes handling eBPF objects in the kernel. eBPF
+object allocations are tracked alongside regular allocations within the cgroup.
+Memory consumption and limits can be queried and set through cgroupfs, the same
+mechanism used for setting memory limits on containers.
+
+## Purpose of package `rlimit`
+
+On kernels supporting memcg accounting, there's no need to manage
+`RLIMIT_MEMLOCK` for effectively using eBPF, as eBPF object allocations now
+count towards the cgroup memory limit instead. However, since many Linux
+distributions still ship pre-5.11 kernels, it's necessary to conditionally
+manage rlimit for kernels lacking memcg accounting for eBPF.
+
+To support writing portable Go tools that work across various kernel versions,
+the `rlimit` package was introduced. It encapsulates two behaviours:
+
+1. As an **import side effect** of importing the package, it lowers the rlimit
+   of the current process to induce a Map creation failure, then restores the
+   original rlimit.
+2. {{ godoc('rlimit/RemoveMemlock') }} conditionally increases `RLIMIT_MEMLOCK`
+   to infinity based on the probe's result. If the kernel supports memcg
+   accounting, this is a no-op.
+
+## Usage
+
+Include this in your application:
+
+{{ go_example('DocRlimit', title="Remove RLIMIT_MEMLOCK if kernel lacks memcg accounting") }}
+
+!!! note ""
+    You can call `RemoveMemlock()` multiple times it your program has
+    multiple entry points or CLI subcommands. The rlimit operation will only
+    execute once.
+
+## Caveats
+
+### Race Conditions
+
+The package was carefully designed with Go's runtime initialization semantics in
```

Comment:
Again, contributor level information. Why do you describe the implementation?

Reply:
Disagree, we've had to explain this to at least one user already. Most of this page is a direct result of https://github.com/cilium/ebpf/issues/1096.

This has the possibility to impact other packages, it's not at all something internal-only.